Remove the restrictive BipedModelType enum. Use the Model Type registry instead.

TagnumElite opened the pull request:
This pull request allows for the use of model types outside of BipedModelType.

The enum BipedModelType ordinal is used to store the citizens model type. This restricts model types to just the enum.
My goal here was to make the IModelTypeRegistry more useful. A new class ModelTypeEntry along withs its Builder allows for
registering custom model types.

Citizens now store there model types instead as a resource location string. This allows the citizen to fetch its model from the model type registry instead of trying to find it in the BipedModelType enum.

Since I 'improved' the IModelTypeRegistry. There is now validation for model types creation.

# Changes proposed in this pull request:
- ~Uses my proposed changes in ldtteam/Structurize#452 to register huts using the HUTS parameter.~
- Change IModeTypeRegistry to support registering ResourceLocations and more to allow for mod model types.
- Renames getModel to getModelType in IJob.

Review please

marchermans commented:
I don't understand this pr

marchermans commented:
This change is completely not needed.
The api gives access to this registry.

Since this is client side only a forge registry is not needed, since their main advantage is the ability to sync between client and server using it's automatic syncing.

TagnumElite commented:
I am currently busy with the change.

Also, the problem with having custom models is them saving them to entity data, it expects the BipedModelType enum.
If I want to have my own models from my own mod, it doesn't save properly.

TagnumElite commented:
I have my course to go to and there is no power so I can only work on this later. I did have an idea for model textures. Instead of specify model texture count and name in code, instead there could be a data generator that generates a file containing a list of textures for both genders for each style. This would allow for validation that all textures exist and allow the modification of texture variations per model, so that it can be changed.

TagnumElite commented:
Sorry, I made the mistake of using the GitHub Client. I had to force overwrite this pull request. To see the old changes look at [this branch](https://github.com/Tagnum-Mods/minecolonies/tree/dedepend_backup).

Justen-N commented:
Tagnum, were you attempting to change the enum at runtime? i.e., to add another type or remove one? If that is the case, then Orion is absolutely right, and you should be able to use the API to add what you were using this for. IN other words, I don't see this as necessary either.

marchermans commented:
The change away from the enum is needed since those are not expandable at runtime. However it suffice to offer a concurrent map via the api and not a full-blown registry

TagnumElite commented:
I am not gonna touch that can of worms, extending enums at runtime. No, the problem I have was just [this](https://github.com/ldtteam/minecolonies/blob/version/main/src/main/java/com/minecolonies/coremod/entity/citizen/citizenhandlers/CitizenJobHandler.java#L88) or more specifically the fact that Citizen EntityData Model was set to enum name and fetched via enum name. All I really did was just change it to a 'ResouceLocation' and fetched everything from the model registry.

Raycoms commented:
> I am not gonna touch that can of worms, extending enums at runtime. No, the problem I have was just [this](https://github.com/ldtteam/minecolonies/blob/version/main/src/main/java/com/minecolonies/coremod/entity/citizen/citizenhandlers/CitizenJobHandler.java#L88) or more specifically the fact that Citizen EntityData Model was set to enum name and fetched via enum name. All I really did was just change it to a 'ResouceLocation' and fetched everything from the model registry.

Okay, I gave this another look. I understand the problem, but I think the solution is an overkill. Instead of changing everything around it, we could just make the BipedModelType a class instead of an enum and add a way to register additional ones.

TagnumElite committed:
Change BipedModelType to class and implement more things

TagnumElite commented:
I can simplify this much further, infact. [Here, I just did exactly that](https://github.com/Tagnum-Mods/minecolonies/commit/41bca7b79a5ac30a6f7fe2aa70315e18c33ab00d).
Is this more in line with what you want?

Raycoms commented:
> I can simplify this much further, infact. [Here, I just did exactly that](https://github.com/Tagnum-Mods/minecolonies/commit/41bca7b79a5ac30a6f7fe2aa70315e18c33ab00d). Is this more in line with what you want?

That looks much better already.

TagnumElite commented:
Well, I can push this branch to that commit or close and make a new pull request if you prefer that commit more.

Raycoms commented:
> Well, I can push this branch to that commit or close and make a new pull request if you prefer that commit more.

Yes please

TagnumElite commented:
Sorry about the above close and re-open. GitHub automatically closed the commit becuase I push before I pulled by mistake.

Raycoms commented on src/api/java/com/minecolonies/api/client/render/modeltype/ModModelTypes.java:
```diff
@@ -0,0 +1,84 @@
+    public static final ResourceLocation PLANTER_ID = new ResourceLocation(Constants.MOD_ID, "planter");
+    public static final ResourceLocation BEEKEEPER_ID = new ResourceLocation(Constants.MOD_ID, "beekeeper");
+
+    public static BipedModelType settler;
```
all upper case, constants

TagnumElite replied:
I copied the format from [ModBuildings.java](https://github.com/ldtteam/minecolonies/blob/version/main/src/api/java/com/minecolonies/api/colony/buildings/ModBuildings.java).

Raycoms commented on src/main/java/com/minecolonies/apiimp/initializer/ModModelTypeInitializer.java:
```diff
@@ -0,0 +1,124 @@
+import com.minecolonies.api.client.render.modeltype.registry.IModelTypeRegistry;
+import com.minecolonies.coremod.client.model.*;
+
+public class ModModelTypeInitializer
```
wrong method formatting in file

TagnumElite replied:
Do you mean you want everything instantiated first, then registered?

Raycoms replied:
No
{

}

TagnumElite replied:
Oh, completely flew over my head. Yeah I will fix that now

TagnumElite committed:
Fix formatting and capitalize constants

marchermans commented on src/api/java/com/minecolonies/api/client/render/modeltype/IModelType.java:
```diff
@@ -23,4 +23,18 @@
+     *
+     * @return The female model
+     */
+    CitizenModel<AbstractEntityCitizen> getFemaleMap();
```
Wrong method name.

marchermans commented on src/api/java/com/minecolonies/api/client/render/modeltype/BipedModelType.java:
```diff
@@ -1,78 +1,72 @@
  * Enum with possible citizens.
  */
-public enum BipedModelType implements ISimpleModelType
+public class BipedModelType implements ISimpleModelType
```
This should stay an enum.

You should never need this in the API anyway.... everwhere where it is in use in the API or internally it should use the ISimpleModelType interface.

marchermans commented on src/api/java/com/minecolonies/api/client/render/modeltype/BipedModelType.java:
```diff
@@ -1,78 +1,72 @@
 package com.minecolonies.api.client.render.modeltype;
```
Who put this in the API? Did I? Because this is an internal implementation detail, and should never be here in the first palce.

marchermans commented on src/api/java/com/minecolonies/api/client/render/modeltype/ModModelTypes.java:
```diff
@@ -0,0 +1,84 @@
+    public static BipedModelType CITIZEN;
+    public static BipedModelType NOBLE;
+    public static BipedModelType ARISTOCRAT;
+    public static BipedModelType BUILDER;
+    public static BipedModelType DELIVERYMAN;
+    public static BipedModelType MINER;
+    public static BipedModelType LUMBERJACK;
+    public static BipedModelType FARMER;
+    public static BipedModelType FISHERMAN;
+    public static BipedModelType UNDERTAKER;
+    public static BipedModelType ARCHER_GUARD;
+    public static BipedModelType KNIGHT_GUARD;
+    public static BipedModelType BAKER;
+    public static BipedModelType SHEEP_FARMER;
+    public static BipedModelType COW_FARMER;
+    public static BipedModelType PIG_FARMER;
+    public static BipedModelType CHICKEN_FARMER;
+    public static BipedModelType COMPOSTER;
+    public static BipedModelType SMELTER;
+    public static BipedModelType COOK;
+    public static BipedModelType STUDENT;
+    public static BipedModelType CRAFTER;
+    public static BipedModelType BLACKSMITH;
+    public static BipedModelType CHILD;
+    public static BipedModelType HEALER;
+    public static BipedModelType TEACHER;
+    public static BipedModelType GLASSBLOWER;
+    public static BipedModelType DYER;
+    public static BipedModelType MECHANIST;
+    public static BipedModelType FLETCHER;
+    public static BipedModelType CONCRETE_MIXER;
+    public static BipedModelType RABBIT_HERDER;
+    public static BipedModelType PLANTER;
+    public static BipedModelType BEEKEEPER;
+
+    private ModModelTypes()
+    {
+        throw new IllegalStateException("Tried to initialize: ModModelTypes but this is a Utility class.");
+    }
+}
```
Again not the way this should be implemented.
You are exposing that BipedModelType which is completely not needed at all.

If this needs to be extendable, make a registry (any will do, even a `ConcurrentHashMap` and have everybody use that) That way you only need to expose the IDs as constants and not the underlying objects, reducing the chance for binary compatibility breaking bugs later on.

TagnumElite replied:
Hmm, I never thought of that. Will make changes shortly.

marchermans commented on src/api/java/com/minecolonies/api/client/render/modeltype/registry/IModelTypeRegistry.java:
```diff
@@ -1,23 +1,32 @@
+     * @param modelType
+     * @return
+     */
+    void register(IModelType modelType);
```
See this is way better.
If you have a registry, why are you exposing the implemented object then?

marchermans reviewed (changes requested):
This now both exposes the API and the underlying implementing class.
That last thing has to go.

Reduce the exposure of the BipedModelType and you are good.

TagnumElite committed:
Fix IModelType typo for getting the female model

Rename `BipedModelType` to `SimpleModelType` and update its documentation.
Simplify all subclasses of `IJob` and `ModModelTypes` to return `IModelType` instead of BipedModelType~SimpleModelType.
Use `ConcurrentHashMap` instead of `ConcurrentMap`.

TagnumElite commented:
So for `IJob#getModel()` do you want me to return a `ResourceLocation` instead of a `IModelType`.

Raycoms commented:
@TagnumElite is this ready or anything missing?

TagnumElite commented:
Mostly, going through the changes I made, I saw that `IModelTypeRegistry` had empty doc strings.

TagnumElite committed:
Update docs on IModelTypeRegistry

Make `#getModelType` Nullable.

TagnumElite commented:
That should be me done unless there are more changes that need to be done. Also, `gradle\configuration.gradle` has incorrect variables `structurize_version` -> `structurizeVersion` and `jei_version` -> `jeiVersion`.
Which will probably be the reason for failed builds.

TagnumElite committed:
Merge branch 'version/main' into dedepend

marchermans reviewed (approved):
Sorry missed that this one is complete 👍 
I had reviewed it already but somehow it got not recorded, blame my phone.
